pkg/kubectl/cmd/apply/prune/prune.go

```diff
@@ -0,0 +1,230 @@
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+)
+
+type Pruner struct {
+	Mapper               meta.RESTMapper
+	GroupVersionKinds    []*meta.RESTMapping
+	ClientFunc           resource.ClientMapperFunc
+	ClientsetFunc        func() (internalclientset.Interface, error)
+	Selector             string
+	IncludeUninitialized bool
+	Cascade              bool
+	GracePeriod          int
+	DryRun               bool
+	Out                  io.Writer
+	ShortOutput          bool
+}
+
+// Param: gvks array of groupVersionKind passed in --prune-whitelist or empty array if no flag.
+// Returns default groupVersionKind mappings to prune, unless overriden by --prune-whitelist flag.
+func ParseGvks(mapper meta.RESTMapper, gvks []string) ([]*meta.RESTMapping, error) {
+
+	// QUESTION: Do I need to explicitly define if namespaced?
+	// Is correct namespace available from "mapping.Scope.Name()"?
+	// Default group/version/kind to prune
+	pruneGvks := [][]string{
+		{"", "v1", "ConfigMap"},
+		{"", "v1", "Endpoints"},
+		{"", "v1", "Namespace"},
+		{"", "v1", "PersistentVolumeClaim"},
+		{"", "v1", "PersistentVolume"},
+		{"", "v1", "Pod"},
+		{"", "v1", "ReplicationController"},
+		{"", "v1", "Secret"},
+		{"", "v1", "Service"},
+		{"batch", "v1", "Job"},
+		{"extensions", "v1beta1", "DaemonSet"},
+		{"extensions", "v1beta1", "Deployment"},
+		{"extensions", "v1beta1", "Ingress"},
+		{"extensions", "v1beta1", "ReplicaSet"},
+		{"apps", "v1beta1", "StatefulSet"},
+		{"apps", "v1beta1", "Deployment"},
+	}
+
+	// If group/version/kind specified in "prune-whitelist" flag, then use those values
+	if len(gvks) != 0 {
+		var err error
+		pruneGvks, err = parsePruneWhitelist(gvks)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Parse groupVersionKind from passed prune-whitelist flag values
+	var groupVersionKinds []*meta.RESTMapping
+	for _, gvk := range pruneGvks {
+		mapping, err := mapper.RESTMapping(schema.GroupKind{Group: gvk[0], Kind: gvk[2]}, gvk[1])
+		if err != nil {
+			return nil, err
+		}
+		groupVersionKinds = append(groupVersionKinds, mapping)
+	}
+
+	return groupVersionKinds, nil
+}
+
+// Example gvks: "core/v1/endpoints", "apps/v1beta1/deployment"
+// Returns validated gvks as array of arrays of strings
+func parsePruneWhitelist(gvks []string) ([][]string, error) {
+
+	var parsedGvks [][]string
+	for _, groupVersionKind := range gvks {
+		gvk := strings.Split(groupVersionKind, "/")
+		if len(gvk) != 3 {
+			return nil, fmt.Errorf("invalid GroupVersionKind format: %v, please follow <group/version/kind>", groupVersionKind)
+		}
+		if gvk[0] == "core" {
+			gvk[0] = ""
+		}
+		parsedGvks = append(parsedGvks, gvk)
+	}
+	return parsedGvks, nil
+}
+
+// Prune resources not previously visited.
+func (p *Pruner) Prune(visitedNamespaces sets.String, visitedUids sets.String) error {
```

Comment:
Move these arguments to be fields of Pruner.  Provide functions VistedNamespace and VisitedUid